contrib/ros/envs/sawyer_env.py

```diff
@@ -0,0 +1,346 @@
+
+    def update_model_states(self, data):
+        if self._simulated:
+            self.model_states = data
+        else:
+            # TODO(gh/2: Finish the 'update_model_states' function for real robot)
+            pass
+
+    def reward(self, achieved_goal, goal):
+        """
+        Compute the reward for current step.
+        :param achieved_goal: the current gripper's position or object's position in the current training episode.
+        :param goal: the goal of the current training episode, which mostly is the target position of the object or the
+                     position.
+        :return reward: float
+                    if sparse_reward, the reward is -1, else the reward is minus distance from achieved_goal to
+                    our goal. And we have completion bonus for two kinds of types.
+        """
+        d = _goal_distance(achieved_goal, goal)
+        if d < self.distance_threshold:
+            return 88
+        else:
+            if self.sparse_reward:
+                return -1.
+            else:
+                return -d
+
+    # Implementation for rllab env functions
+    # -----------------------------------
+    @rate_limited(100)
+    def step(self, action):
+        """
+        Perform a step in gazebo. When end of episode
+        is reached, reset() should be called to reset the environment's internal state.
+        Input
+        -----
+        action : an action provided by the environment
+        Outputs
+        -------
+        (observation, reward, done, info)
+        observation : agent's observation of the current environment
+        reward [Float] : amount of reward due to the previous action
+        done : a boolean, indicating whether the episode has ended
+        info : a dictionary containing other diagnostic information from the previous action
+        """
+        self._robot.set_command(action)
+
+        obs = self._get_obs()
+
+        reward = self.reward(obs['achieved_goal'], self.goal)
+        done = _goal_distance(obs['achieved_goal'],
+                              self.goal) < self.distance_threshold
+        next_observation = obs['observation'].copy()
+        return Step(observation=next_observation, reward=reward, done=done)
+
+    def reset(self):
+        """
+        Resets the state of the environment, returning an initial observation.
+        Outputs
+        -------
+        observation : the initial observation of the space. (Initial reward is assumed to be 0.)
+        """
+        self.goal = copy.deepcopy(self.sample_goal())
+
+        if self._simulated:
+            self._gazebosrv.set_model_pose(
+                model_name='target',
+                new_pose=Pose(
+                    position=Point(
+                        x=self.goal[0], y=self.goal[1], z=self.goal[2])),
+                reference_frame='world')
+            self._reset_sim()
+        else:
+            # TODO(gh/3: Finish the reset function for real robot)
```

Comment:
```python
# TODO(gh/???): Finish the reset function for real robot
```
Where `???` is the issue number github assigns your todo when you file it here:https://github.com/ryanjulian/rllab/issues